### ISSUE TITLE
AI difficulty - to weak

### ISSUE THREAD

**Issue by FrostBlade5** [wesnoth/wesnoth#6156]
Is there anyone working on the wesnoth-AI ? i noticed the AI can well calculate how chances are to win or lose a single battle between 2 units, but overall strategy from AI is bad compared to human. Is there any way to increase AI difficulty? The experimental AI seems to be even worse. If you look at other games such as chess, AI is even smarter than the best human-being. Ofcourse chess and wesnoth are 2 completly different things.

So im not sure if this can be called an issue, but sort of. Personally i feel its booring to fight the AI.

**Comment by Wedge009** [comment]
Firstly, this is a very vague and subjective request. I suggest reviewing https://github.com/wesnoth/wesnoth/labels/AI and seeing if any of those would cover you - I'd argue this is probably a duplicate of #1136, #1145, #1170, #1293, #1692, #1807, and #2337. The difference being that those are specific in their requests and give concrete goals to work towards.

Secondly, the comparison with chess isn't really helpful. Chess is a well-studied problem with strict rules and a limited game space to work in. While I'm sure algorithms have improved over the years, I suspect the advances in memory and performance allows computers to consider many, many more moves in advance than even the best human players, giving it a significant advantage. Wesnoth also has relatively simple rules but the game space is much large, is non-uniform, and there is also the element of luck to consider. These aren't trivial to work with.

Lastly, our primary AI developer recently stepped down from active maintenance so it's unlikely that this will be worked on any time soon.

While I think people are welcome to and are encouraged to report issues, I'm not so sure it's helpful to create spurious requests like this one as it just creates more work potentially unnecessarily. I apologise if you did attempt to look at existing AI issues logged, but it doesn't appear to be the case here.

**Comment by max-torch** [comment]
We can say "requests about AI should be specific enough to address".

**Comment by mattsc** [comment]
I agree that this is not a useful issue.  If you have specific suggestions/comments such as in the issues @Wedge009 links to above, feel free to add them and reopen (or open new issues, one per suggestion, if there are more than one), but for now I am closing this.

**Comment by cooljeanius** [comment]
Please also look at the wiki pages about AI: 

- https://wiki.wesnoth.org/WhyWritingAWesnothAIIsHard 
- https://wiki.wesnoth.org/Creating_Custom_AIs
- https://wiki.wesnoth.org/Category:AI

**Comment by FrostBlade5** [comment]
Issue 1:
Wesnoth AI does some strange moves, for exmaple ai has a unit with low hp say 10-15 and instead of retreating to save it or attacking to deal some damage before it diesanyway, the ai just walks directly next to my unit, then does nothing.
In the next turn im able to kill him easly and gain XP for free.
A human wouldnt play that way.

Issue 2:
AI is attacking with king instead of saving him which many many times made it easy for me to defeat AI.
The king should have a different AI than normal units, atleast a tiny bit, dont move into stupid positions/situations with King, thats what the AI is doing.

Issue 3:
One of my units had 19 out of 20 XP untill it will level up.
AI attacks my unit and he levels up, the AI took damage and is now low HP while my unit gained a level.
This is stupid behaviour.
I would never attack any unit that only needs 1-2 XP to level up unless the unit is low hp and im 90% sure i have a good chance to kill it before it gains the levelup.

**Comment by FrostBlade5** [comment]
> 
> 
> I agree that this is not a useful issue. If you have specific suggestions/comments such as in the issues @Wedge009 links to above, feel free to add them and reopen (or open new issues, one per suggestion, if there are more than one), but for now I am closing this.

How can i reopen?

**Comment by knyghtmare** [comment]
> Wesnoth AI does some strange moves, for exmaple ai has a unit with low hp say 10-15 and instead of retreating to save it or attacking to deal some damage before it dies anyway, the ai just walks directly next to my unit, then does nothing.
In the next turn im able to kill him easly and gain XP for free.
A human wouldnt play that way.

> AI is attacking with king instead of saving him which many many times made it easy for me to defeat AI.
The king should have a different AI than normal units, atleast a tiny bit, dont move into stupid positions/situations with King, thats what the AI is doing.

> One of my units had 19 out of 20 XP untill it will level up.
AI attacks my unit and he levels up, the AI took damage and is now low HP while my unit gained a level.
This is stupid behaviour.
I would never attack any unit that only needs 1-2 XP to level up unless the unit is low hp and im 90% sure i have a good chance to kill it before it gains the levelup.

All of your problems are not bugs.
They are all solvable via `[ai][/ai]` code inside `[side][/side]` defines and with `[micro_ai][/micro_ai]`

review the documentations on these before bringing up trivial issues.

**Comment by mattsc** [comment]
I would not have phrased it quite the same way as @knyghtmare, but yes, those are all known characteristics of the AI.  The Wesnoth AI needs to be able to play all maps/settings somehow, and a pretty aggressive default setting has been found to be the best compromise for that.  But there is no one size fits all, so this will inevitably result in some undesirable default behavior, which is why the AI is very configurable.  @knyghtmare gave some examples of how that can be done.  Would it be nice for the AI to adapt its behavior automatically?  Of course, but so far nobody knows how to do that so that it results in more than an incremental improvement.

Also, I could bring counter examples to all your points, but will only do so for the last: attacking high-XP units was actually specifically added to the AI as beforehand it was possible to put such units at strategic points on the map and make the AI entirely passive.  So which is the correct way?  The answer is that it depends.  So the default setting is one that works better on average (most players think that playing against a somewhat over-aggressive AI is more fun than against one that is overly passive) and a scenario/map author can change it to fit that scenario/map.  [FTR, I would probably use a different solution than the one suggested by knyght (I'd remove the candidate action that does these attacks), but the answer to which solution to choose is again: it depends.]

So no, this issue should not be reopened based on what you wrote.  Writing an AI for Wesnoth is difficult.  We'll likely never get one that is as good as a human player.  That's not an argument for not trying to improve it, but everything you have written so far is both known and has known solutions.

**Comment by FrostBlade5** [comment]
!bump

Once again i have to comment on AI behaviour...
The Wesnoth AI is sort of agressive and in general it is a decent enemy, however there are some scenarios where the AI is honestly speaking, just dumb.
Yesterday i played a scenario, i had about 10 elven-horse-riders and i moved all of them to the edge of the map, then i went straight up to the enemys king and killed it, what's dumb about this? Well the enemy had more than enough units to prevent me from taking the king out but instead of blocking my units on the edge of the map, the AI only attacked like 2-3 of them and the rest passed through and reached the king. A real player would never allow this to happen.

Generally speaking the AI seems to lack a tactic/technique/strategy.

I complained about the AI of "Kings" before, when i started the match yesterday there was multiple AI enemys and one of them managed to get it's king killed right at the beginning of the game. A real player would protect his king at all costs and not send it to death. I can already smell to upcomming comments, get over it, blabla... Honestly it's just annoying that the AI sends their king to die and protect's it badly. Personally i would have programmed the AI to set a priority on protecting the king as soon as an enemy unit is in attack range to it. All in all, even tho this might never happen i will say it anyways because it has to be said, the king should have it's own AI and not use the same AI as all other units, the king should have a priority to stay alive as long as possible and not leave the castle to get killed straight away, period.

I would also like to ask, is it possible to collect anonymous statistic data from wesnoth games to improve AI?
Or, is it possible to let the wesnoth AI fight against it self over and over again to become smarter and smarter untill eventually it will outsmart the average player? Or atleast add difficultys such as easy / medium / hard for the AI to be selected at the start of the game? If you are familiar with alpha zero and alpha go, the AI that beat a human who once was world champion, basically that's what im asking if such a "self learning AI" would be possible for wesnoth aswell?
The current AI is not bad but it could be much better, and by better i mean more challenging which would make the game all in all more fun.
